### PR TITLE
Implemented Korach Moran Zaks for asynchronous clique

### PR DESCRIPTION
**Pull request by Wasik93**
Added test
Added example
Added implementation of Korach-Moran-Zaks
Added algo description
Changed makefile

### CODE REVIEW

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+	}
+
+ASK_RESPONSE:
+	received_index, received_message := receiveMessage(node)
+	message_type := received_message.Flag
+
+	switch message_type {
+	case ASK:
+		if state.Phase > received_message.Phase || (state.Phase == received_message.Phase && state.King > received_message.King) {
+			goto ASK_RESPONSE
```
Comment (krzysztof-turowski) [inline leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go:80]: Please re-write it as a for loop from L73 on with continue here

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	new_state.Queue = make([]message, 0)
+	setState(node, &new_state)
+}
+
+func run(node lib.Node) {
+	node.StartProcessing()
+	initialize(node)
+	new_state := getState(node)
+	new_state.Phase = 0
+	setState(node, &new_state)
```
Comment (krzysztof-turowski) [inline leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go:314]: I don't get it, in initialize (L301) you already set Phase = 0

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+	mainKing(node)
```
Comment (krzysztof-turowski) [inline leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go:315]: Please refactor it in the way that there is more-or-less one for loop with every iteration: receive message, read state, process, send x messages to other nodes, finish iteration

In general, a structure like run() in leader/undirected_ring/async_itai_rodeh/async_itai_rodeh.go would be appreciated.

File: text/1180971/async_korach_moran_zaks.tex [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    \item ASK - a king can send at most len(King.Vassals) + 1, therefore all the kings in this phase sent together at most n messages, the total number of messages sent by kings is bounded by \(n\log_2(n)\), Every citizen transfers at most one ASK message per phase, the total number of such messages sent by all citizens is also bounded by \(n\log_2(n)\) 
+    \item ACCEPT - Citizens send no more than ASK messages - bounded by \(n\log_2(n)\), Kings send exactly \(n-1\) messages
+    \item UPDATE - one such message per phase - bounded by  \(n\log_2(n)\) 
+\end{enumerate}
+    So the total number of messages is bounded by \(5n\log_2(n) + 2n\). This bound is pretty loose. The algorithm uses around 800 messages for n = 100 but the bound is equal to 3200.
+\end{proof}
+
+\begin{thebibliography}{9}
+\bibitem{article}
+Ephraim none Korach, Shlomo Moran, Schmuel Zaks (1984)
```
Comment (krzysztof-turowski) [inline text/1180971/async_korach_moran_zaks.tex:210]: It's rather not "Ephraim none Korach", please correct

File: text/1180971/async_korach_moran_zaks.tex [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+If \(king_j\) directly receives the message he immediately become the citizen. Otherwise it will be transferred by Vassals to the king. The message could be blocked on the way to \(king_j\) but that means that the node that blocked the message is waiting for UPDATE from his king as a response for ASK message with higher status than status(\(king_j\)). It means that one of the messages will cause the 
+\(king_j\) to become the citizen.
+\end{proof}
+By Lemma 3 we get contradiction to the assumptions that \(s > 1\)
+The algorithm will stop after the King of Kings announce his presence (by sending the I\_AM\_THE\_KING message down the tree).
+
+\section*{Proof of complexity}
+\begin{theorem}
+The number of messages sent is \(5n\log_2(n) + \mathcal{O}(n)\)
+\end{theorem}
```
Comment (krzysztof-turowski) [inline text/1180971/async_korach_moran_zaks.tex:186]: Move it after Lemma 4

**Review by krzysztof-turowski — COMMENTED** [review]
In general the code is very good, I'd suggest refactoring only for the purposes of exhibiting better the structure "receive-decodeState-process-send-saveState"

File: text/1180971/async_korach_moran_zaks.tex (hunk 5 above)
Comment (Wasik93) [reply text/1180971/async_korach_moran_zaks.tex]: Done

File: text/1180971/async_korach_moran_zaks.tex (hunk 4 above)
Comment (Wasik93) [reply text/1180971/async_korach_moran_zaks.tex]: Done

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go (hunk 3 above)
Comment (Wasik93) [reply leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go]: Done

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go (hunk 2 above)
Comment (Wasik93) [reply leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go]: It was removed

File: leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go (hunk 1 above)
Comment (Wasik93) [reply leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go]: It is no longer needed to have goto or for loop here

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much for your contribution

**Pull request merged**